File: src/test/java/io/github/alexengrig/nih/collections/ArrayTest.java
package io.github.alexengrig.nih.collections;

import org.junit.Test;

import static org.junit.Assert.*;

public class ArrayTest {
    @Test
    public void should_create_newArray_by_length() {
        final int length = 10;
        final Array<Object> array = new Array<>(length);
        assertEquals("Array length", length, array.length());
        assertEquals("Array size", 0, array.size());
        assertTrue("Array must be empty", array.isEmpty());
        assertFalse("Array must be full", array.isFull());
    }

    @Test
    public void should_create_newArray_by_values() {
        final Object[] expected = new Object[]{1, 2, 3};
        final int length = expected.length;
        final Array<Object> array = new Array<>(expected);
        assertEquals("Array length", length, array.length());
        assertEquals("Array size", length, array.size());
        assertFalse("Array must be empty", array.isEmpty());
        assertTrue("Array must be full", array.isFull());
    }

    @Test
    public void should_create_newArray_by_varargs() {
        final Object[] expected = new Object[]{1, 2, 3};
        final int length = expected.length;
        final Array<Object> array = new Array<>(expected[0], expected[1], expected[2]);
        assertEquals("Array length", length, array.length());
        assertEquals("Array size", length, array.size());
        assertFalse("Array must be empty", array.isEmpty());
        assertTrue("Array must be full", array.isFull());
    }

    @Test
    public void should_add_value_to_array() {
        final int length = 1;
        final Array<Object> array = new Array<>(length);
        array.add(new Object());
        assertEquals("Array length", length, array.length());
        assertEquals("Array size", length, array.size());
        assertFalse("Array must be empty", array.isEmpty());
        assertTrue("Array must be full", array.isFull());
    }

    @Test
    public void should_add_values_to_array() {
        final Object[] values = new Object[]{new Object(), new Object()};
        final int length = values.length;
        final Array<Object> array = new Array<>(length);
        array.addAll(values);
        assertEquals("Array length", length, array.length());
        assertEquals("Array size", length, array.size());
        assertFalse("Array must be empty", array.isEmpty());
        assertTrue("Array must be full", array.isFull());
    }

    @Test
    public void should_return_value_from_array() {
        final Object expected = new Object();
        final Array<Object> array = new Array<>(new Object(), expected);
        assertEquals("Element at index 1", expected, array.get(1));
    }

    @Test
    public void should_remove_value_from_array() {
        final Object expected = new Object();
        final Array<Object> array = new Array<>(expected);
        assertEquals("Array length", 1, array.length());
        assertEquals("Array size", 1, array.size());
        assertFalse("Array must be empty", array.isEmpty());
        assertTrue("Array must be full", array.isFull());
        final Object actual = array.remove(0);
        assertEquals("Array length", 1, array.length());
        assertEquals("Array size", 0, array.size());
        assertTrue("Array must be empty", array.isEmpty());
        assertFalse("Array must be full", array.isFull());
        assertEquals("Removed element", expected, actual);
    }

    @Test
    public void should_remove_firstValue_from_array() {
        final Object expected = new Object();
        final Array<Object> array = new Array<>(expected);
        assertEquals("Array length", 1, array.length());
        assertEquals("Array size", 1, array.size());
        assertFalse("Array must be empty", array.isEmpty());
        assertTrue("Array must be full", array.isFull());
        final Object actual = array.removeFirst();
        assertEquals("Array length", 1, array.length());
        assertEquals("Array size", 0, array.size());
        assertTrue("Array must be empty", array.isEmpty());
        assertFalse("Array must be full", array.isFull());
        assertEquals("Removed element", expected, actual);
    }

    @Test
    public void should_remove_lastValue_from_array() {
        final Object expected = new Object();
        final Array<Object> array = new Array<>(expected);
        assertEquals("Array length", 1, array.length());
        assertEquals("Array size", 1, array.size());
        assertFalse("Array must be empty", array.isEmpty());
        assertTrue("Array must be full", array.isFull());
        final Object actual = array.removeLast();
        assertEquals("Array length", 1, array.length());
        assertEquals("Array size", 0, array.size());
        assertTrue("Array must be empty", array.isEmpty());
        assertFalse("Array must be full", array.isFull());
        assertEquals("Removed element", expected, actual);
    }
}

File: src/test/java/io/github/alexengrig/nih/collections/ArrayTest.kt
package io.github.alexengrig.nih.collections

import org.junit.Assert
import org.junit.Test

class ArrayTest {
    @Test
    fun should_create_newArray_by_length() {
        val length = 10
        val array = Array<Any>(length)
        Assert.assertEquals("Array length", length.toLong(), array.length().toLong())
        Assert.assertEquals("Array size", 0, array.size().toLong())
        Assert.assertTrue("Array must be empty", array.isEmpty)
        Assert.assertFalse("Array must be full", array.isFull)
    }

    @Test
    fun should_create_newArray_by_values() {
        val expected = arrayOf<Any>(1, 2, 3)
        val length = expected.size
        val array = Array(expected)
        Assert.assertEquals("Array length", length.toLong(), array.length().toLong())
        Assert.assertEquals("Array size", length.toLong(), array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
    }

    @Test
    fun should_create_newArray_by_varargs() {
        val expected = arrayOf<Any>(1, 2, 3)
        val length = expected.size
        val array = Array(expected[0], expected[1], expected[2])
        Assert.assertEquals("Array length", length.toLong(), array.length().toLong())
        Assert.assertEquals("Array size", length.toLong(), array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
    }

    @Test
    fun should_add_value_to_array() {
        val length = 1
        val array = Array<Any>(length)
        array.add(Any())
        Assert.assertEquals("Array length", length.toLong(), array.length().toLong())
        Assert.assertEquals("Array size", length.toLong(), array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
    }

    @Test
    fun should_add_values_to_array() {
        val values = arrayOf(Any(), Any())
        val length = values.size
        val array = Array<Any>(length)
        array.addAll(values)
        Assert.assertEquals("Array length", length.toLong(), array.length().toLong())
        Assert.assertEquals("Array size", length.toLong(), array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
    }

    @Test
    fun should_return_value_by_index_from_array() {
        val expected = Any()
        val array = Array(Any(), expected)
        Assert.assertEquals("Element at index 1", expected, array[1])
    }

    @Test
    fun should_return_firstValue_from_array() {
        val expected = Any()
        val array = Array(expected, Any())
        Assert.assertEquals("First element", expected, array.first)
    }

    @Test
    fun should_return_lastValue_from_array() {
        val expected = Any()
        val array = Array(Any(), expected)
        Assert.assertEquals("Last element", expected, array.last)
    }

    @Test
    fun should_remove_value_from_singletonArray() {
        val expected = Any()
        val array = Array(expected)
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 1, array.length().toLong())
        Assert.assertEquals("Array size", 1, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
        val actual = array.remove(0)
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 1, array.length().toLong())
        Assert.assertEquals("Array size", 0, array.size().toLong())
        Assert.assertTrue("Array must be empty", array.isEmpty)
        Assert.assertFalse("Array must be full", array.isFull)
        Assert.assertEquals("Removed element", expected, actual)
    }

    @Test
    fun should_remove_value_from_doubleArray() {
        val expected = Any()
        val array = Array(expected, Any())
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 2, array.length().toLong())
        Assert.assertEquals("Array size", 2, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
        val actual = array.remove(0)
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 1, array.end.toLong())
        Assert.assertEquals("Array length", 2, array.length().toLong())
        Assert.assertEquals("Array size", 1, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertFalse("Array must be full", array.isFull)
        Assert.assertEquals("Removed element", expected, actual)
    }

    @Test
    fun should_remove_firstValue_from_singletonArray() {
        val expected = Any()
        val array = Array(expected)
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 1, array.length().toLong())
        Assert.assertEquals("Array size", 1, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
        val actual = array.removeFirst()
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 1, array.length().toLong())
        Assert.assertEquals("Array size", 0, array.size().toLong())
        Assert.assertTrue("Array must be empty", array.isEmpty)
        Assert.assertFalse("Array must be full", array.isFull)
        Assert.assertEquals("Removed element", expected, actual)
    }

    @Test
    fun should_remove_firstValue_from_doubleArray() {
        val expected = Any()
        val array = Array(expected, Any())
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 2, array.length().toLong())
        Assert.assertEquals("Array size", 2, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
        val actual = array.removeFirst()
        Assert.assertEquals("Begin", 1, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 2, array.length().toLong())
        Assert.assertEquals("Array size", 1, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertFalse("Array must be full", array.isFull)
        Assert.assertEquals("Removed element", expected, actual)
    }

    @Test
    fun should_remove_lastValue_from_singletonArray() {
        val expected = Any()
        val array = Array(expected)
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 1, array.length().toLong())
        Assert.assertEquals("Array size", 1, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
        val actual = array.removeLast()
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 1, array.length().toLong())
        Assert.assertEquals("Array size", 0, array.size().toLong())
        Assert.assertTrue("Array must be empty", array.isEmpty)
        Assert.assertFalse("Array must be full", array.isFull)
        Assert.assertEquals("Removed element", expected, actual)
    }

    @Test
    fun should_remove_lastValue_from_doubleArray() {
        val expected = Any()
        val array = Array(Any(), expected)
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 0, array.end.toLong())
        Assert.assertEquals("Array length", 2, array.length().toLong())
        Assert.assertEquals("Array size", 2, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertTrue("Array must be full", array.isFull)
        val actual = array.removeLast()
        Assert.assertEquals("Begin", 0, array.begin.toLong())
        Assert.assertEquals("End", 1, array.end.toLong())
        Assert.assertEquals("Array length", 2, array.length().toLong())
        Assert.assertEquals("Array size", 1, array.size().toLong())
        Assert.assertFalse("Array must be empty", array.isEmpty)
        Assert.assertFalse("Array must be full", array.isFull)
        Assert.assertEquals("Removed element", expected, actual)
    }
}